lib/rucio/daemons/rsedecommissioner/profiles/generic.py

```diff
@@ -0,0 +1,477 @@
+from rucio.core.lock import get_replica_locks, get_replica_locks_for_rule_id
+from rucio.core.replica import list_replicas_per_rse, set_tombstone, update_replica_state
+from rucio.core.request import get_request_by_did
+from rucio.core.rse import add_rse_attribute, update_rse
+from rucio.core.rule import list_rules_for_rse_decommissioning, move_rule, update_rule
+from rucio.db.sqla.constants import ReplicaState
+
+from ..config import DecommissioningStatus, set_status
+from ..throttling import DeletionLimitExceeded, LimitChecker
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, Iterable
+    from typing import Any
+
+
+def generic_profile_template_function(
+    rse: 'dict[str, Any]',
+    config: 'dict[str, Any]',
+    initializer: 'Callable[..., None]',
+    discoverer: 'Callable[..., Iterable[dict[str, Any]]]',
+    handlers: 'list[tuple[Callable[..., bool], Callable[..., bool]]]',
+    finalizer: 'Callable[..., None]',
+    *,
+    logger: 'Callable[..., None]' = logging.log
+) -> None:
+    """Generic decommissioning template function.
+
+    Decommissioning under a generic profile proceeds in common steps of
+    - Profile initialization
+    - Discovery of rules to act (move / delete / etc.) on
+    - Actual handling of the rules
+    - Finalization
+    This function takes the functions corresponding to the four steps as its arguments and
+    executes them in order.
+
+    The handlers are given as a list of pairs of functions. The two functions both have a
+    signature
+    (rule, rse, *, logger=logging.log) -> bool
+    with the following actions:
+    - The first function is the "condition" against which each rule is tested. The returned
+      boolean indicates whether the rule satisfies it.
+    - The second function is the "(in)action" taken against the rule if the first function
+      returns True. The returned boolean indicates whether the limit checker should be
+      incremented after executing the action.
+
+    For each rule listed by the discoverer function, the conditions are applied in the order
+    given in the handlers list until one evaluates to True, at which point the corresponding
+    action function is executed. All remaining condition-action pairs are skipped for the rule.
+
+    :param rse: RSE table entry as a dictionary.
+    :param config: Decommissioning configuration dictionary.
+    :param initializer: Profile initialization function.
+    :param discoverer: Function to find and list the rules to process.
+    :param handlers: A list of (condition, action) functions.
+    :param finalizer: Finalization function.
+    :param logger: Logging function.
+    """
+    max_locks = config_get_int('rse-decommissioner', 'max_locks', default=-1)
+    max_rules = config_get_int('rse-decommissioner', 'max_rules', default=-1)
+    if max_locks < 0 and max_rules < 0:
+        raise RuntimeError('Throttling required for a non-dry run')
```

Comment:
Does a dry-run mode exist?

Reply:
Changed the error message to "This daemon requires configuration parameters rse-decommissioner/max_locks and rse-decommissioner/max_rules set to non-negative integers."